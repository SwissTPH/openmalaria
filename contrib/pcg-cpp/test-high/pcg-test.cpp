/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
