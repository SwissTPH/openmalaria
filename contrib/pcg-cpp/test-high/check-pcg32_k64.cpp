#define RNG pcg32_k64
#define TWO_ARG_INIT 1

#include "pcg-test.cpp"

