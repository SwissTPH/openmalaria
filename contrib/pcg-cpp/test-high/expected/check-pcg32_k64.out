pcg32_k64:
      -  result:      32-bit unsigned int
      -  period:      2^2112   (* 2^63 streams)
      -  size:        272 bytes

Round 1:
  32bit: 0xe85244a0 0x7112822f 0x9325f975 0xf50dea01 0x8cec9bba 0xaa9fa4b3
  Again: 0xe85244a0 0x7112822f 0x9325f975 0xf50dea01 0x8cec9bba 0xaa9fa4b3
  Coins: HHTTTHTHTTHHHHHHTHTHTHTTHHTHTHHHHTTTHHHTTHTHHHHTHTTHTTTTTTTTHTTTT
  Rolls: 4 6 3 6 2 3 3 5 5 2 4 3 6 6 2 5 3 6 1 3 1 2 6 4 6 5 4 5 6 1 3 2 2
   -->   rolling dice used 33 random numbers
  Cards: Qs 2d 6s 8c 6d 4d 5d Ah Ks 3d 5s Qd 6h Td 5c 4h 2c 8h Ad 2s Js 7c
	 5h Qc Ac 7s 4s Jd Kc 9c 9s 2h Kh Ts Qh Jc 7h 8d Th Tc 3h As Jh 4c
	 8s 3c 9h 6c 3s 7d 9d Kd

Round 2:
  32bit: 0x98c55d02 0x0fd39f52 0x11b56e6c 0x54cbf14e 0x45bde3a6 0xa77ab20f
  Again: 0x98c55d02 0x0fd39f52 0x11b56e6c 0x54cbf14e 0x45bde3a6 0xa77ab20f
  Coins: THTHTHTHTTHHTHTHHTTTHTTHTTHTHTTHTTHTTTHHTHHTTTTHTHHTTTHHTTTTTTHHH
  Rolls: 5 2 1 3 5 5 3 5 4 2 4 1 2 1 1 6 6 6 6 1 1 2 1 4 5 1 1 1 6 4 1 2 6
   -->   rolling dice used 33 random numbers
  Cards: 9s 6s 7c 4h 4d 8d 8h Jc 3h Qh 5s 9c 3s Tc 8s As Kh Kc 2c 2s Ac 8c
	 5c 4c 6c 2h Jh 5d 9d Qs 7d Qc Th Ks Ts 3c Js 7s 6h 7h Kd 5h Qd 9h
	 3d Td Ad Ah 2d 4s 6d Jd

Round 3:
  32bit: 0xf2fe5c00 0xacc7f4f5 0xeac130c7 0x7a2ce2be 0xd39bf181 0xce784be4
  Again: 0xf2fe5c00 0xacc7f4f5 0xeac130c7 0x7a2ce2be 0xd39bf181 0xce784be4
  Coins: THTTHTHTTHTTHHTHTHHHTHTTHTHTHHTHHTHTHHTTTHHHHTHTHTTHHTHTHTHHTTTTH
  Rolls: 4 2 3 3 2 6 6 1 6 1 4 4 3 4 1 2 2 1 5 6 1 1 2 6 5 4 4 6 3 1 5 1 1
   -->   rolling dice used 33 random numbers
  Cards: 7c Jc 9s Ah 4h 3d 2d 5d 3s 7h Kh 6s 6h Ks 2c 9c 2s 8s Tc 7d 5h 3c
	 8h 4d Jh Td 4s 5s 2h Ac 9d 6c Js Ts Qd 8c 5c Ad 7s Kd Qc Jd Th 9h
	 Qh 6d 8d Kc Qs 3h As 4c

Round 4:
  32bit: 0x1ef66795 0x83d37b1c 0xfc6d9564 0x9806e724 0xe10fbb53 0xdc1d6eab
  Again: 0x1ef66795 0x83d37b1c 0xfc6d9564 0x9806e724 0xe10fbb53 0xdc1d6eab
  Coins: TTTTHHHHHTTTTHHTHTTTHHTHTHTHHTHTTHTHTHTHHHHHTHHTHTHTTTHHTTTTHHHTT
  Rolls: 4 2 2 4 6 6 5 2 4 1 1 6 6 3 3 5 3 3 2 3 3 6 6 5 2 6 6 1 5 6 2 6 4
   -->   rolling dice used 33 random numbers
  Cards: 2d 4s Qs 8h Jd 6d Jh 2s 7c 3s 5h Kd Js Td 3c 9s As Ts Ah Tc 8d 7h
	 5s 9h Th 7d 4c 8c Qc 5d 5c Qh Qd 8s Ac 6c 3d 2h 4h Kc Jc 9c 4d 6s
	 Ks 7s Kh 9d Ad 6h 2c 3h

Round 5:
  32bit: 0x24b9177f 0x83dde908 0xd2719639 0x43d1afbf 0x35db4e5a 0x5d9f5718
  Again: 0x24b9177f 0x83dde908 0xd2719639 0x43d1afbf 0x35db4e5a 0x5d9f5718
  Coins: HHHHTHHHTTHTTTHHTTTTTTHTTHTHHHTTTHTHHHTTTTHTHTTTTTHTHHHHHHHHTTHTT
  Rolls: 1 6 1 6 5 1 6 1 1 1 1 6 3 6 4 4 3 6 3 3 3 5 6 4 4 4 3 2 3 2 4 5 1
   -->   rolling dice used 33 random numbers
  Cards: 6c Th Kc Qd 7d 6s Ks 7h Tc 9c 5s 9h 8d 9d 7s 5d Kh Ad 6d Qc Td 9s
	 Js 5h 6h Qh 2s Ts 2h Jd 3c Ac Qs 8s 5c 2c As 2d 7c 4d 3s 8c Kd 4c
	 4s Jc Ah Jh 3h 4h 8h 3d

